mlflow/tensorflow/__init__.py

```diff
@@ -887,6 +968,110 @@ def _log_early_stop_callback_metrics(callback, history, metrics_logger):
         if metric_key is not None:
             metrics_logger.record_metrics(restored_metrics, stopped_epoch + 1)
 
+    def _log_keras_model_with_warning(history, args):
+        try:
+            if log_models:
+                import types
+
+                def _get_input_data_slice():
+                    input_training_data_original = args[0]
+                    input_example_slice = None
+
+                    def _extract_from_sequence_or_generator():
+                        input_training_data = input_training_data_original
+                        if isinstance(input_training_data_original, types.GeneratorType):
+                            input_training_data = next(input_training_data_original)[0]
+                        elif isinstance(
+                            input_training_data_original, tensorflow.keras.utils.Sequence
+                        ):
+                            input_training_data = input_training_data_original[:][0]
+                        return input_training_data
+
+                    input_training_data = _extract_from_sequence_or_generator()
+                    if isinstance(input_training_data, np.ndarray):
+                        input_example_slice = input_training_data[:INPUT_EXAMPLE_SAMPLE_ROWS]
+                    elif isinstance(input_training_data, tensorflow.data.Dataset):
+                        steps = 1
+                        if history.params is not None and "steps" in history.params:
+                            steps = history.params["steps"]
+
+                        def _extract_n_steps(input_example_n_steps):
+                            if steps > 1:
+                                return np.array(
+                                    [
+                                        v[0][:INPUT_EXAMPLE_SAMPLE_ROWS]
+                                        for v in input_example_n_steps.take(1).as_numpy_iterator()
+                                    ]
+                                )[0]
+                            return np.array(
+                                [
+                                    v[0]
+                                    for v in input_example_n_steps.take(
+                                        INPUT_EXAMPLE_SAMPLE_ROWS
+                                    ).as_numpy_iterator()
+                                ]
+                            )
+
+                        input_example_slice = _extract_n_steps(input_training_data)
```

Comment:
Not fully following what's going on here. Why does extracting the first `INPUT_EXAMPLE_SAMPLE_ROWS` from a `tf.Dataset` depend on `history.params["steps"]`? Is this step-dependent logic necessary? If it is, can we add a clear docstring to this method to explain why?

Reply:
will add an example for it (please refer to this (unit test)[https://github.com/mlflow/mlflow/blob/2d57dd8ae6a12368b118dcaa25b7d925a8582956/tests/tensorflow/test_tensorflow2_autolog.py#L1388]), but if the dataset is batched, then this just extracts the first batch and then next extracts the first 5 samples from that first batch. Refer to: (steps_per_epoch)[https://www.tensorflow.org/api_docs/python/tf/keras/Model#fit]